mlir/lib/Gradient/Utils/GradientShape.cpp

```diff
@@ -113,17 +113,16 @@ std::vector<Type> computeBackpropTypes(func::FuncOp callee,
 
         Type diffArgType = fnType.getInput(diffArgIndices[i]);
 
-        if (auto tensorType = diffArgType.dyn_cast<RankedTensorType>()) {
+        if (auto tensorType = dyn_cast<RankedTensorType>(diffArgType)) {
             ArrayRef<int64_t> tensorShape = tensorType.getShape();
             diffArgType = tensorType.getElementType();
```

Comment:
Should we assert that the element type is a float (like below)?

Reply:
Sure, I'll add a utility method that recursively verifies shaped types.

Reply:
I don't think it needs to be recursive, unless you are saying we can tensors of tensors?